### PR TITLE
Adjust array resizing in block builder

### PR DESCRIPTION
**Pull request by dnhatn**
I looked into an async profiler and found that `AbstractBlockBuilder#updatePosition` was consuming a significant amount of CPU. This is because we're growing the `firstValueIndexes` array one by one. While this minimizes wasted memory, it requires more CPU. I think we should use `ArrayUtil.oversize()` to resize this array.

In contrast, it appears that we're growing the values array too quickly by 50% each time. I think we should use `ArrayUtil.oversize()` with a growth rate of 1/8 here as well. 

However, it's fine if we prefer to keep the current resizing for the `values` array since we compact this array when building the block. But I think we should change the resizing strategy for the `firstValueIndexes` array.

### CODE REVIEW

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractBlockBuilder.java [hunk 1]
```diff
@@ -173,8 +169,10 @@ protected void adjustBreaker(long deltaBytes) {
     private void setFirstValue(int position, int value) {
         if (position >= firstValueIndexes.length) {
             final int currentSize = firstValueIndexes.length;
-            adjustBreaker((long) (position + 1 - currentSize) * Integer.BYTES);
-            firstValueIndexes = Arrays.copyOf(firstValueIndexes, position + 1);
```
Comment (dnhatn) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractBlockBuilder.java:177]: We're growing the `firstValueIndexes` array one by one.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractBlockBuilder.java [hunk 2]
```diff
@@ -159,11 +160,6 @@ public final void close() {
      */
     protected void extraClose() {}
 
-    static int calculateNewArraySize(int currentSize) {
-        // trivially, grows array by 50%
-        return currentSize + (currentSize >> 1);
```
Comment (dnhatn) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractBlockBuilder.java:164]: We're growing the `values` array by 50% each time.

**Comment by dnhatn** [comment]
update-position

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/X-BlockBuilder.java.st [hunk 3]
```diff
@@ -327,28 +327,19 @@ $endif$
...
                 theBlock = blockFactory.newConstant$Type$BlockWith(values[0], 1, estimatedBytes);
+            } else if (estimatedBytes > blockFactory.maxPrimitiveArrayBytes()) {
+                theBlock = buildBigArraysBlock();
+            } else if (isDense() && singleValued()) {
+                theBlock = blockFactory.new$Type$ArrayVector(values, positionCount, estimatedBytes).asBlock();
             } else {
-                if (estimatedBytes > blockFactory.maxPrimitiveArrayBytes()) {
-                    theBlock = buildBigArraysBlock();
-                } else {
-                    if (values.length - valueCount > 1024 || valueCount < (values.length / 2)) {
```
Comment (dnhatn) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/X-BlockBuilder.java.st:334]: We don't need to compact if we grow the `values` array by 1/8 each time.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractBlockBuilder.java [hunk 4]
```diff
@@ -173,8 +169,10 @@ protected void adjustBreaker(long deltaBytes) {
     private void setFirstValue(int position, int value) {
         if (position >= firstValueIndexes.length) {
             final int currentSize = firstValueIndexes.length;
-            adjustBreaker((long) (position + 1 - currentSize) * Integer.BYTES);
-            firstValueIndexes = Arrays.copyOf(firstValueIndexes, position + 1);
+            final int newLength = ArrayUtil.oversize(position + 1, Integer.BYTES);
+            adjustBreaker((long) newLength * Integer.BYTES);
+            firstValueIndexes = ArrayUtil.growExact(firstValueIndexes, newLength);
+            adjustBreaker(-(long) currentSize * Integer.BYTES);
```
Comment (nik9000) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractBlockBuilder.java:175]: It's worth saying that we could grow the `values` array at the same rate as `firstValues` but that'd be a bit more complex. And maybe less efficient. But maybe more? I have no idea. Either way, this is fine. But you did make a choice to get here.

**Review by nik9000 — APPROVED** [review]

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractBlockBuilder.java (hunk 4 above)
Comment (dnhatn) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractBlockBuilder.java:176]: I added a comment in https://github.com/elastic/elasticsearch/pull/106934/commits/bc12f69ae759f2a5b32b8d10cf2bd46de902fa81

**Pull request merged**